test/lifecycle_test.dart

```diff
@@ -175,6 +177,52 @@ main() {
 
         sharedTypeTests(testSnapshotType);
       });
+
+      group('getDerivedStateFromProps returns', () {
+        test('the returned state updates properly', () {
+          const Map initialState = const {
+            'initialState': 'initial',
+          };
+
+          final Map initialProps = unmodifiableMap({'getDerivedStateFromProps': (_, __, ___) => initialState});
+          final Map expectedProps = unmodifiableMap(defaultProps, initialProps, emptyChildrenProps);
+
+          LifecycleTestHelper component = getDartComponent(render(components2.LifecycleTest(initialProps)));
+
+          component.lifecycleCalls.clear();
+
+          (component as react.Component2).forceUpdate();
+
+          expect(
+              component.lifecycleCalls,
+              equals([
+                matchCall('getDerivedStateFromProps'),
+                matchCall('render', state: initialState),
+                matchCall('getSnapshotBeforeUpdate', args: [expectedProps, initialState], state: initialState),
+                matchCall('componentDidUpdate', args: [expectedProps, initialState, null], state: initialState),
+              ].where((matcher) => matcher != null).toList()));
+        });
+
+        test('null properly', () {
+          final Map initialProps = unmodifiableMap({'getDerivedStateFromProps': (_, __, ___) => null});
+          final Map expectedProps = unmodifiableMap(defaultProps, initialProps, emptyChildrenProps);
+
+          LifecycleTestHelper component = getDartComponent(render(components2.LifecycleTest(initialProps)));
+
+          component.lifecycleCalls.clear();
+
+          (component as react.Component2).forceUpdate();
+
+          expect(
+              component.lifecycleCalls,
+              equals([
+                matchCall('getDerivedStateFromProps'),
+                matchCall('render', state: {}),
```

Comment:
Can we also add an arbitrary state key/value pair to this component so that we can assert that returning null from getDerivedStateFromProps does not result in the unexpected mutation of state -- that the pre-existing key/value pairs are not obliterated?